### PR TITLE
KAFKA-7538: Reduce lock contention for Partition ISR lock

### PR DESCRIPTION
**Pull request by rajinisivaram**
Check for ISR updates using ISR read lock and acquire ISR write lock only if ISR needs to be updated. This avoids lock contention between request handler threads processing log appends on the leader holding the ISR read lock and request handler threads processing replica fetch requests that check/update ISR.

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

### CODE REVIEW

**Comment by rajinisivaram** [comment]
Test metrics from the test included in the PR:
**PartitionLockTest#testNoLockContentionWithoutIsrUpdate: delays: append=300, check=10, update=50**
(note that these are unrealistic delays introduced to create a stable automated test)
```
Old code:
  Metrics: appendAvg=315.73 appendMax=318.65 replicaFetchUpdateAvg=309.72 replicaFetchUpdateMax=326.00
  Time taken for 300 records with 2 fetchers and 3 producers : 97087 ms.

New code:
  Metrics: appendAvg=304.50 appendMax=328.19 replicaFetchUpdateAvg=11.67 replicaFetchUpdateMax=14.60
  Time taken for 300 records with 2 fetchers and 3 producers : 30525 ms.
```
**PartitionLockTest#testNoLockContentionWithoutIsrUpdate: delays: append=0, check=0, update=0:**
(the test results vary between runs here since there are no delays, but average numbers are close in the two cases, but the maximum is consistently better with the new code)

```
Old code:
Time taken for 3000000 records with 2 fetchers and 3 producers : 68884 ms.
Metrics: appendAvg=0.035 appendMax=13.993 replicaFetchUpdateAvg=0.017 replicaFetchUpdateMax=13.997

New code:
Time taken for 3000000 records with 2 fetchers and 3 producers : 48904 ms.
Metrics: appendAvg=0.033  appendMax=11.72 replicaFetchUpdateAvg=0.011 replicaFetchUpdateMax=11.138
```

File: core/src/main/scala/kafka/cluster/Partition.scala [hunk 1]
```diff
@@ -497,30 +512,51 @@ class Partition(val topicPartition: TopicPartition,
    * @return true if the high watermark has been updated
    */
   def maybeExpandIsr(replicaId: Int, logReadResult: LogReadResult): Boolean = {
-    inWriteLock(leaderIsrUpdateLock) {
+    var hwIncremented = false
+    val (leaderReplica, needsIsrUpdate) = inReadLock(leaderIsrUpdateLock) {
```
Comment (ijuma) [inline core/src/main/scala/kafka/cluster/Partition.scala]: Could we use `StampedLock.asReadLock` or something like that for better performance?

File: core/src/main/scala/kafka/cluster/Partition.scala [hunk 2]
```diff
@@ -578,10 +614,14 @@ class Partition(val topicPartition: TopicPartition,
...
+   *
    * Note There is no need to acquire the leaderIsrUpdate lock here
-   * since all callers of this private API acquire that lock
+   * since all callers of this private API acquire either ISR read or write lock.
+   * HW update is performed while holding the Partition object lock with
+   * locking order ISR lock -> Partition object lock. So this method should not
+   * acquire an ISR read or write lock.
    */
-  private def maybeIncrementLeaderHW(leaderReplica: Replica, curTime: Long = time.milliseconds): Boolean = {
+  private def maybeIncrementLeaderHW(leaderReplica: Replica, curTime: Long = time.milliseconds): Boolean = synchronized {
```
Comment (hachikuji) [inline core/src/main/scala/kafka/cluster/Partition.scala]: Looking at trunk, it seems like we no longer need this synchronization. We basically do the same thing by relying on the lock in `Log` when updating the high watermark. As far as I can tell, this is the only thing we really need to change in this patch.

Comment (rajinisivaram) [reply core/src/main/scala/kafka/cluster/Partition.scala]: @hachikuji Thanks for the review,  I have rebased on trunk and removed this synchronization.

File: core/src/main/scala/kafka/cluster/Partition.scala (hunk 1 above)
Comment (rajinisivaram) [reply core/src/main/scala/kafka/cluster/Partition.scala]: @ijuma Thanks for the review, Since StampedLocks are not reentrant, the code may become harder to maintain. So I left the lock as ReentrantReadWriteLock, even though as you say it doesn't perform as well for this case. Since updates are relatively infrequent, perhaps that is ok?

**Comment by rajinisivaram** [comment]
@hachikuji @ijuma Thanks for the review. I have rebased the PR on trunk. There is overlap between this PR and @bdbyrne 's PR https://github.com/apache/kafka/pull/7921. I haven't reviewed that yet, but perhaps we could merge the two.

**Comment by hachikuji** [comment]
@rajinisivaram I think we should probably just merge this one and close #7921. I will review this morning.

File: core/src/main/scala/kafka/cluster/Partition.scala [hunk 3]
```diff
@@ -850,6 +871,16 @@ class Partition(val topicPartition: TopicPartition,
       tryCompleteDelayedRequests()
   }
 
+  private[cluster] def needsShrinkIsr(): Boolean = {
+    leaderLogIfLocal match {
```
Comment (hachikuji) [inline core/src/main/scala/kafka/cluster/Partition.scala]: nit: we could probably use `isLeader` since we don't need the log.

File: core/src/test/scala/unit/kafka/cluster/PartitionLockTest.scala [hunk 4]
```diff
@@ -0,0 +1,369 @@
...
+            append(partition, numRecordsPerProducer, followerQueues)
+          } catch {
+            case e: Throwable =>
+              PartitionLockTest.this.error("Exception during append", e)
+              throw e
+          }
+        }
+      })
+    })
+    (1 to numReplicaFetchers).foreach(i => {
```
Comment (hachikuji) [inline core/src/test/scala/unit/kafka/cluster/PartitionLockTest.scala]: nit: `foreach { i =>`?

File: core/src/test/scala/unit/kafka/cluster/PartitionLockTest.scala [hunk 5]
```diff
@@ -0,0 +1,369 @@
...
+          Thread.sleep(isrUpdateDelayMs)
+        super.maybeUpdateIsrAndVersion(isr, zkVersionOpt)
+      }
+
+      override def createLog(replicaId: Int, isNew: Boolean, isFutureReplica: Boolean, offsetCheckpoints: OffsetCheckpoints): Log = {
+        val log = super.createLog(replicaId, isNew, isFutureReplica, offsetCheckpoints)
+        new SlowLog(log, mockTime, appendDelayMs)
+      }
+
+      override def inReadLock[T](lock: ReadWriteLock)(fun: => T): T = {
```
Comment (hachikuji) [inline core/src/test/scala/unit/kafka/cluster/PartitionLockTest.scala]: This feels a bit low level to me. Presumably a different locking implementation could get the same benefit of avoiding contention. Do you think they add much value to these test case?

File: core/src/test/scala/unit/kafka/cluster/PartitionLockTest.scala [hunk 6]
```diff
@@ -0,0 +1,369 @@
...
+
+  val numReplicaFetchers = 2
+  val numProducers = 3
+  val numRecordsPerProducer = 5
+  // Use artificially large delays to enable automated testing of locking delays
+  val appendDelayMs = 300
+  val isrCheckDelayMs = 10
+  val isrUpdateDelayMs = 50
+
+  val mockTime = new MockTime()
```
Comment (hachikuji) [inline core/src/test/scala/unit/kafka/cluster/PartitionLockTest.scala]: The mix of mock time and system time in these tests is a little hard to follow. Maybe a brief comment of explanation would help.

File: core/src/test/scala/unit/kafka/cluster/PartitionLockTest.scala [hunk 7]
```diff
@@ -0,0 +1,369 @@
...
+
+  private def concurrentProduceFetch(numProducers: Int, numReplicaFetchers: Int, numRecords: Int, appendDelayMs: Long): Unit = {
+    val followerQueues = (0 until numReplicaFetchers).map(_ => new ArrayBlockingQueue[MemoryRecords](2))
+
+    val totalRecords = numRecordsPerProducer * numProducers
+    val futures = new ArrayBuffer[Future[_]]()
+
+    (0 until numProducers).foreach(_ => {
+      futures += executorService.submit(new Runnable {
+        override def run(): Unit = {
```
Comment (hachikuji) [inline core/src/test/scala/unit/kafka/cluster/PartitionLockTest.scala]: I think now we can use the native scala lambdas. Instead of `new Runnable`, we can do `() =>`

**Review by hachikuji — COMMENTED** [review]
Thanks, this looks good. Left a few small comments.

File: core/src/test/scala/unit/kafka/cluster/PartitionLockTest.scala (hunk 6 above)
Comment (rajinisivaram) [reply core/src/test/scala/unit/kafka/cluster/PartitionLockTest.scala]: The tests were originally written to get some timings. I converted that to an integration test, but actually, that is probably not that useful to detect regressions since the times used in the test need to be large enough to avoid flaky tests. I have rewritten the tests to use semaphores to verify the conditions without relying on actual time. Hopefully the code is easier to understand and also more reliable.

File: core/src/test/scala/unit/kafka/cluster/PartitionLockTest.scala (hunk 5 above)
Comment (rajinisivaram) [reply core/src/test/scala/unit/kafka/cluster/PartitionLockTest.scala]: Removed this and added semaphores for testing.

**Comment by rajinisivaram** [comment]
@hachikuji Thanks for the review. I have addressed the comments and also rewritten the test to avoid reliance on run time,

File: core/src/main/scala/kafka/cluster/Partition.scala [hunk 8]
```diff
@@ -850,6 +872,15 @@ class Partition(val topicPartition: TopicPartition,
       tryCompleteDelayedRequests()
   }
 
+  private[cluster] def needsShrinkIsr(): Boolean = {
+    if (isLeader) {
+        val outOfSyncReplicaIds = getOutOfSyncReplicas(replicaLagTimeMaxMs)
```
Comment (hachikuji) [inline core/src/main/scala/kafka/cluster/Partition.scala]: nit: fix indentation

File: core/src/main/scala/kafka/cluster/Partition.scala [hunk 9]
```diff
@@ -173,6 +173,18 @@ case class SimpleAssignmentState(replicas: Seq[Int]) extends AssignmentState
...
+ * 1) Partition is thread-safe. Operations on partitions may be invoked concurrently from different
+ *    request handler threads
+ * 2) ISR updates are synchronized using a read-write lock. Read lock is used to check if an update
+ *    is required to avoid acquiring write lock in the common case of replica fetch when no update
+ *    is performed. ISR update condition is checked a second time under write lock before performing
+ *    the update
+ * 3) Various other operations like leader changes are processed while holding the ISR write lock.
+ *    This can introduce delays in produce and replica fetch requests, but these operations are typically
+ *    infrequent.
+ * 4) HW updates are synchronized using ISR read lock.
```
Comment (hachikuji) [inline core/src/main/scala/kafka/cluster/Partition.scala]: Should we mention the locking in `Log`?

File: core/src/test/scala/unit/kafka/cluster/PartitionLockTest.scala [hunk 10]
```diff
@@ -0,0 +1,292 @@
...
+    }
+
+    // Release sufficient append permits to complete all except one append. Verify that
+    // follower state update completes even though an update is in progress. Then release
+    // the permit for the final append and verify follower update for the last append.
+    appendSemaphore.release(numProducers * numRecordsPerProducer - 1)
+
+    // If a semaphore that triggers contention is present, verify that state update futures
+    // haven't completed. Then release the semaphore to enable all threads to continue.
+    contentionSemaphore match {
```
Comment (hachikuji) [inline core/src/test/scala/unit/kafka/cluster/PartitionLockTest.scala]: nit: `foreach`?

File: core/src/main/scala/kafka/cluster/Partition.scala [hunk 11]
```diff
@@ -814,7 +833,10 @@ class Partition(val topicPartition: TopicPartition,
   private def tryCompleteDelayedRequests(): Unit = delayedOperations.checkAndCompleteAll()
 
   def maybeShrinkIsr(): Unit = {
-    val leaderHWIncremented = inWriteLock(leaderIsrUpdateLock) {
+    val needsIsrUpdate = inReadLock(leaderIsrUpdateLock) {
```
Comment (hachikuji) [inline core/src/main/scala/kafka/cluster/Partition.scala:837]: Not a major issue, but since `needsShrinkIsr` is exposed, maybe we can move the lock acquire to it?

File: core/src/test/scala/unit/kafka/cluster/PartitionLockTest.scala [hunk 12]
```diff
@@ -0,0 +1,292 @@
...
+    TestUtils.waitUntilTrue(() => shrinkIsrSemaphore.hasQueuedThreads, "shrinkIsr not invoked")
+    concurrentProduceFetch(numProducers, numReplicaFetchers, numRecordsPerProducer, Some(shrinkIsrSemaphore))
+    active.set(false)
+    future.get(15, TimeUnit.SECONDS)
+  }
+
+  private def concurrentProduceFetch(numProducers: Int,
+                                     numReplicaFetchers: Int,
+                                     numRecords: Int,
+                                     contentionSemaphore: Option[Semaphore]): Unit = {
```
Comment (hachikuji) [inline core/src/test/scala/unit/kafka/cluster/PartitionLockTest.scala]: nit: would be nice to use a consistent mechanism to propagate semaphores to threads. Also, should this be `shrinkIsrSemaphore`?

**Review by hachikuji — COMMENTED** [review]
Thanks @rajinisivaram . Just a few more small comments. The test case is looking good.

File: core/src/main/scala/kafka/cluster/Partition.scala (hunk 11 above)
Comment (rajinisivaram) [reply core/src/main/scala/kafka/cluster/Partition.scala:837]: Actually `needsShrinkIsr` can be private, updated. Left the lock in here to be consistent with `needsExpandIsr`.

**Comment by rajinisivaram** [comment]
@hachikuji Thanks for the review, have addressed the comments.

**Review by hachikuji — APPROVED** [review]
LGTM. Thanks for the PR!

**Comment by rajinisivaram** [comment]
retest this please

**Comment by rajinisivaram** [comment]
@hachikuji Thanks for the review, merging to trunk. Streams test failure is not related.
